### PR TITLE
Tunable CCL global settings at the Python-level

### PR DESCRIPTION
**Pull request by nikfilippas**
This PR decouples the (otherwise connected at the C level) CCL accuracy parameters from `Cosmology` objects, at the Python level.

In a new module `parameters.py` I have implemented the class `CCLParameters` which wraps the SWIG-generated classes for the spline parameters, the GSL parameters, and the physical constants.

In the spirit of making `Cosmology` objects truly immutable (which they ought to be) I have deprecated the option to mutate the accuracy parameters of any instantiated `Cosmology`: all new instances are populated with the global parameters at the time they are created.

Users can access (and set) the global parameters using `ccl.gsl_params`, `ccl.spline_params`, and `ccl.physical_constants`.

To do this I removed the `const` type definition of the C-level spline and GSL parameters, so they are added in `ccllib.cvar` by SWIG. Physical constants were already mutable (because `T_CMB` is temporarily mutated in `Cosmology`).

This feature comes with the advantage that we can remove all dummy cosmologies from the codebase. Every time we need one just to look up the accuracy parameters, we can just peek at the parameters at the time of instantiation, if a `Cosmology` is not provided.

Maybe it's also worth considering having something like `ccl.spline_params.set_accuracy("high" | "medium" | "low")` with pre-made parameter sets. High accuracy would be what is expected from LSST-like analyses with slower computation, while low accuracy will give a considerable increase in speed. This is one of the things mentioned in #588.

### CODE REVIEW

**Comment by tilmantroester** [comment]
@nikfilippas Do you want to rebase/merge master and I then rewrite the warning to match the new `gsl_params` setup?

**Comment by nikfilippas** [comment]
@tilmantroester okay, merged and adjusted the warning as required. Have you managed to have a look? Would you like to review this?

File: pyccl/errors.py
Comment (damonge) [inline pyccl/errors.py:45]: Why do we need this in this PR? Is this a leftover from the v3 PRs?

File: pyccl/pk2d.py
Comment (damonge) [inline pyccl/pk2d.py:61]: Is this still the case? (I.e. is this still used to determine sampling rates?)

File: pyccl/tests/test_cclerror.py
Comment (damonge) [inline pyccl/tests/test_cclerror.py:65]: Should all these live in a different PR?

File: src/ccl_core.c
Comment (damonge) [inline src/ccl_core.c:632]: Since you created these two nice functions, should we put them to use in `src/ccl_power.c` to reduce the clutter there?

File: pyccl/parameters.py
Comment (damonge) [inline pyccl/parameters.py:49]: same

File: pyccl/parameters.py
Comment (damonge) [inline pyccl/parameters.py:52]: just explain this to me

**Review by damonge — CHANGES_REQUESTED** [review]
OK, thanks a lot @nikfilippas . First review. Main thing is you need to explain to me this `object` business you're using in parameters.py

File: pyccl/__init__.py
Comment (nikfilippas) [reply pyccl/__init__.py:73]: Yes, in the spirit of making Cosmology objects truly immutable, I am now allowing `spline_params` and `gsl_params` of a particular Cosmology to change, but a `CCLDeprecationWarning` will be thrown.
I think this is good enough for the time being until v3, where the deprecation cycle will come to an end and trying to change any params after instantiation will throw an error.

File: pyccl/boltzmann.py
Comment (nikfilippas) [reply pyccl/boltzmann.py:122]: uhmm I have no opinion on that. I just copied what was already there. I suggest let's leave it as is for now.

File: src/ccl_core.c
Comment (nikfilippas) [reply src/ccl_core.c:632]: These are very similar to the 2 functions below it. Instead of calculating `na` and `nk` from `cosmo`, they calculate `na` and `nk` from a `ccl_parameters` struct. So, all 4 functions belong to the same file. Shall they live in `ccl_core.c` or in `ccl_power.c`? Your call.

File: pyccl/parameters.py
Comment (damonge) [inline pyccl/parameters.py:38]: Add a comment that this is what will happen in v3 (i.e. an error will be thrown)

File: benchmarks/test_tracers.py
Comment (damonge) [reply benchmarks/test_tracers.py:23]: OK, I see. I'd like to get rid of this altogether. It seems extremely silly. I'll open an issue.

File: pyccl/pk2d.py
Comment (damonge) [reply pyccl/pk2d.py:61]: Can you add "Will be deprecated in the next CCL version"?

File: pyccl/parameters.py
Comment (damonge) [inline pyccl/parameters.py:49]: Just so I understand: here you're using `object.__setattr__` instead of `self._bak` because you changed `self`'s setattr below, right?

File: src/ccl_core.c
Comment (damonge) [reply src/ccl_core.c:632]: No, no, they can live here or wherever you want. What I mean is why not use them in `ccl_power.c` to generate the k and a arrays? E.g. [here](https://github.com/LSSTDESC/CCL/blob/c01d3b032bff936cc752a7f2b6b94e23758d9d9d/src/ccl_power.c#L59) or [here](https://github.com/LSSTDESC/CCL/blob/c01d3b032bff936cc752a7f2b6b94e23758d9d9d/src/ccl_power.c#L76).

**Review by damonge — COMMENTED** [review]
Ok, a couple of new comments and one old one remaining.

One overall comment: in an ideal world I'd have preferred to start by cleaning up the C layer so actually none of this need to live there. This would simplify a lot of the code you've had to add here, but c'est la vie. I think this is what we'll do in the future.

File: pyccl/pk2d.py
Comment (nikfilippas) [reply pyccl/pk2d.py:61]: Let's not deprecate this. `cosmo` is still used to extrapolate backwards in time using the linear power spectrum, when the requested scale factor is outside of the range of the `Pk2D` object.
With my changes, `eval` will work as before, but users won't be able to extrapolate backwards.

Comment (nikfilippas) [reply pyccl/pk2d.py:61]: Oh wait no! I got my PRs mixed up. No, we won't be deprecating this. `cosmo` will stay there, but I think I've addressed your point in another PR of mine, #923 .

File: pyccl/parameters.py
Comment (nikfilippas) [reply pyccl/parameters.py:38]: Done.

File: pyccl/parameters.py
Comment (nikfilippas) [reply pyccl/parameters.py:49]: Yes. Before `__init__` is called, `self`'s dictionary doesn't contain the entry `_bak`. So if you try to do `self._bak` directly, it will `raise KeyError(f"Parameter _bak does not exist.")`.

File: src/ccl_core.c
Comment (nikfilippas) [reply src/ccl_core.c:632]: Oh, you said "put them to use in", I read "put them in".

Comment (nikfilippas) [reply src/ccl_core.c:632]: Okay, I know what you mean but I can't see which lines should be replaced exactly. And why would they have to be replaced with my functions that accept a `ccl_parameters` struct? The functions you showed me take `cosmo` as an argument. The C code already has versions of my functions that accept `cosmo`. So why bother with my functions?

**Comment by nikfilippas** [comment]
@damonge back to you. I agree with your overall comment. CCL has to be re-written in C++ like @tilmantroester has suggested, but changing the entire internal structure will require a lot of work. This PR does the job in what I believe is an elegant way, but at some point how CCL handles the parameters needs to change.

File: pyccl/ccl_core.i
Comment (tilmantroester) [inline pyccl/ccl_core.i:9]: Can't we remove that now and disallow changing `A_SPLINE_MAX` in the `__setattribute__` method?

File: pyccl/parameters.py
Comment (tilmantroester) [inline pyccl/parameters.py:66]: `A_SPLINE_MAX` should be immutable.

File: pyccl/ccl_core.i
Comment (nikfilippas) [reply pyccl/ccl_core.i:9]: Yes we can, but I didn't want to touch the C-code for that one in case someone changes the Python-code in the future and forgets to reinstate that check. There are a couple of places this check appears in the C-code.

File: pyccl/parameters.py
Comment (nikfilippas) [reply pyccl/parameters.py:66]: It is. Here, for example I am making use of the C-code. If `A_SLINE_MAX` gets set to 1, then the function returns doing nothing. If not, the C-code takes over and raises.

**Comment by damonge** [comment]
@tilmantroester @nikfilippas are we happy to leave C-level modifications for a future PR and merge this one? If so, I can approve.

File: pyccl/parameters.py
Comment (tilmantroester) [reply pyccl/parameters.py:66]: Could you make it raise an exception even when being set to 1? `A_MAX_SPLINE` shouldn't be touched, even when trying to assign the "correct" value.

Comment (nikfilippas) [reply pyccl/parameters.py:66]: I would argue it's okay to set it to its correct value. It is actually touched (set) whenever the parameters are reloaded to their default values, through `spline_params.reload()`.

Also in the (near) future, we might want to relax this requirement to see how things evolve at scale factors larger than 1 (i.e. the far future).

Comment (nikfilippas) [reply pyccl/parameters.py:66]: Jokes aside, if we have it raise an error, I would have to fall back to `object.__setattr__` in `reload`, which is slightly slower.

Comment (tilmantroester) [reply pyccl/parameters.py:66]: Fine. I'd just caution that the C code implicitly assumes `A_MAX_SPLINE==1` in a couple of places and allowing setting `A_MAX_SPLINE` needs a careful review of much of the C code base to make sure this is okay.

**Comment by nikfilippas** [comment]
@damonge I think there is consensus now.

**Comment by damonge** [comment]
@nikfilippas can you resolve the conflicts? I'll approve then when the tests pass.

**Comment by nikfilippas** [comment]
@damonge Done.

**Review by damonge — APPROVED** [review]
LGTM!

**Pull request merged**
